services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManagerImpl.java

```diff
@@ -650,6 +652,62 @@ void buildTableIndexWithLock(final ProgressCallback progressCallback, final IdAn
 		}
 	}
 	
+	void attemptToRestoreTableFromExistingSnapshot(IdAndVersion idAndVersion, String tableResetToken, long targetChangeNumber) {
+		// If there are any changes that are already applied to the table we let it build as normal
+		if (tableIndexDao.getMaxCurrentCompleteVersionForTable(idAndVersion) > -1L) {
+			return;
+		}
+		
+		tableManagerSupport.getMostRecentTableSnapshot(idAndVersion).ifPresent( snapshot -> {
+			IdAndVersion snapshotId = IdAndVersion.newBuilder().setId(snapshot.getTableId()).setVersion(snapshot.getVersion()).build();
+			
+			// Keeps the change number that correspond with the snapshot
+			long snapshotChangeNumber = tableManagerSupport.getLastTableChangeNumber(snapshotId).orElseThrow(
+				() -> new IllegalStateException("Expected a change number for snapshot " + snapshotId + ", but found none.")
+			);
+			
+			log.info("Restoring table " + idAndVersion + " from snapshot " + snapshotId + "...");
+			
+			tableManagerSupport.attemptToUpdateTableProgress(
+				idAndVersion, 
+				tableResetToken, 
+				"Restoring table " + idAndVersion + " from snapshot " + snapshotId, 
+				snapshotChangeNumber, 
+				targetChangeNumber
+			);
+			
+			// Clear the table and restore the snapshot including the table indices
+			deleteTableIndex(idAndVersion);
+
+			IndexDescription indexDescription = new TableIndexDescription(idAndVersion);
+			List<ColumnModel> snapshotSchema = tableManagerSupport.getTableSchema(snapshotId);
+			
+			// Setup the table with the schema of the snapshot
+			setIndexSchema(indexDescription, snapshotSchema);
+			
+			boolean isSearchEnabled = tableManagerSupport.isTableSearchEnabled(snapshotId);
+			// Makes sure to sync the search flag with the version from the snapshot
+			setSearchEnabled(idAndVersion, isSearchEnabled);
+			
+			// Restore the table data from the snapshot
+			tableManagerSupport.restoreTableFromS3(idAndVersion, snapshot.getBucket(), snapshot.getKey());
+			
+			// Optimize the table
+			optimizeTableIndices(idAndVersion);

```

Comment:
maybe also a common method that does:

1. optimization
2. populate index 
3. refresh seach